File: core/src/nl/t64/game/rpg/screens/questlog/QuestListListener.java
package nl.t64.game.rpg.screens.questlog;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.InputListener;
import com.badlogic.gdx.scenes.scene2d.ui.List;
import lombok.AllArgsConstructor;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.audio.AudioCommand;
import nl.t64.game.rpg.audio.AudioEvent;
import nl.t64.game.rpg.components.quest.QuestGraph;

import java.util.Optional;
import java.util.function.Consumer;


@AllArgsConstructor
class QuestListListener extends InputListener {

    private final List<QuestGraph> questList;
    private final Consumer<QuestGraph> populateQuestSpecifics;

    @Override
    public boolean keyDown(InputEvent event, int keycode) {
        return switch (keycode) {
            case Input.Keys.UP,
                    Input.Keys.DOWN,
                    Input.Keys.HOME,
                    Input.Keys.END -> isSelected();
            case Input.Keys.LEFT -> setSelectedUp();
            case Input.Keys.RIGHT -> setSelectedDown();
            default -> false;
        };
    }

    private boolean setSelectedUp() {
        questList.setSelectedIndex(Math.max(questList.getSelectedIndex() - 10, 0));
        return isSelected();
    }

    private boolean setSelectedDown() {
        questList.setSelectedIndex(Math.min(questList.getSelectedIndex() + 10, questList.getItems().size - 1));
        return isSelected();
    }

    private boolean isSelected() {
        return Optional.ofNullable(questList.getSelected())
                       .map(this::isSelected)
                       .orElse(false);
    }

    private boolean isSelected(QuestGraph quest) {
        Utils.getAudioManager().handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_MENU_CURSOR);
        populateQuestSpecifics.accept(quest);
        return true;
    }

}

File: core/src/nl/t64/game/rpg/screens/questlog/QuestListListener.kt
package nl.t64.game.rpg.screens.questlog

import com.badlogic.gdx.Input
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.InputListener
import com.badlogic.gdx.scenes.scene2d.ui.List
import nl.t64.game.rpg.Utils.audioManager
import nl.t64.game.rpg.audio.AudioCommand
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.components.quest.QuestGraph
import nl.t64.game.rpg.constants.Constant
import kotlin.math.max
import kotlin.math.min


private const val TEN = 10

internal class QuestListListener(
    private val questList: List<QuestGraph>,
    private val populateQuestSpecifics: (QuestGraph) -> Unit
) : InputListener() {

    override fun keyDown(event: InputEvent, keycode: Int): Boolean {
        return when (keycode) {
            Input.Keys.UP, Input.Keys.DOWN, Input.Keys.HOME, Input.Keys.END -> isSelected()
            Constant.KEYCODE_L1, Input.Keys.LEFT -> setSelectedUp()
            Constant.KEYCODE_R1, Input.Keys.RIGHT -> setSelectedDown()
            else -> false
        }
    }

    private fun setSelectedUp(): Boolean {
        if (questList.items.isEmpty) {
            questList.setSelectedIndex(max(questList.selectedIndex - TEN, -1))
        } else {
            questList.setSelectedIndex(max(questList.selectedIndex - TEN, 0))
        }
        return isSelected()
    }

    private fun setSelectedDown(): Boolean {
        questList.selectedIndex = min(questList.selectedIndex + TEN, questList.items.size - 1)
        return isSelected()
    }

    private fun isSelected(): Boolean {
        return questList.selected?.let { isSelected(it) } ?: false
    }

    private fun isSelected(quest: QuestGraph): Boolean {
        audioManager.handle(AudioCommand.SE_PLAY_ONCE, AudioEvent.SE_MENU_CURSOR)
        populateQuestSpecifics.invoke(quest)
        return true
    }

}
